Python 3 Centos5 fixes

jeanconn opened the pull request:
This
  - installs llvmlite at the last version compiled on CentOS5 (0.18.0)
  - adds an actual "pinned" file to pin it there
  - adds a system .condarc to get mostly not-CentOS 6 packages from the anaconda channels
  - adds sherpa 4.9.1 from the cxc dev channel (channel controlled via that dot_condarc file)

It looks like pinning llvmlite does not restrict Python to 3.6.0, as if I try to update Python, we get 3.6.2.  I had thought Python would just float up to that version if available and not pinned on the original install, but am guessing that Python itself is handled differently in this regard.

jeanconn commented:
We could put a `python>=3.6.1` or 3.6.2 line in pkgs.conda if you want a "non-upgrade" install of the newer version of Python in ska3/flight .

jeanconn commented:
I'm also not sure about the best way now to evaluate changes to confirm that the set I pick has the least impact on `py3` to `master` diffs.

taldcroft commented:
Why is the llvmlite 0.18.0 spec in two places?  Which one wins if there is a conflict?

taldcroft commented:
Have you confirmed that the Python binary is an "old" (CentOS-5) build?  Is it possible (in reasonable time) to check that all packages which are installed are from CentOS-5 builds?

jeanconn commented:
> Why is the llvmlite 0.18.0 spec in two places? Which one wins if there is a conflict?

Good point.  The pin file pins the version and the pkgs.conda file independently asks for the package to actually be installed.  I suppose in pkgs.conda there shouldn't really need to be a requested version and it can live only in the comment (since this is our first pinned package).

jeanconn commented:
> Have you confirmed that the Python binary is an "old" (CentOS-5) build? Is it possible (in reasonable time) to check that all packages which are installed are from CentOS-5 builds?

No.  I'll get the 3.6.1 and 3.6.2 binaries and see if I can figure that out.  I was hoping that we just didn't have to care if the some module and ska_testr tests passed.

 The fundamental problem is that generally we have no real insight into what built the package until something doesn't work.  And I don't know what things to test to try to find if something doesn't work.  For numba this is easy, because its tests don't pass on a CentOS 5 system llvmlite is > 0.18.0 .  For other packages, our thread with anaconda suggested that newer packages shouldn't end up in the old channels (which are now restricted by the `.condarc`) but some newer packages are still ending up in the old channels *and* there is no guarantee that a package will use a version of conda new enough properly document it is a new build in the burned in recipe (and if the conda is new enough it has the right debug info... the package is probably built on the new compilers).

I had started to just try to verify that all `.so` files were ABI compatible but was hoping we could skip that if things just basically seemed to work in ska_testr.

jeanconn commented:
I can't confirm that the last Python 3.6.2 in the pkgs/free channel was built on CentOS5, but I can confirm the contrapositive... that the first Python 3.6.2 in pkgs/main channel was definitely built on the new compilers and explicitly doesn't work.

```Package plan for installation in environment /proj/sot/ska3/flight/arch/x86_64-linux_CentOS-5:

The following NEW packages will be INSTALLED:

    libedit:      3.1-heed3624_0       
    libgcc-ng:    7.2.0-h7cc24e2_2     
    libstdcxx-ng: 7.2.0-h7a57d05_2     
    ncurses:      6.0-h9df7e31_2       

The following packages will be UPDATED:

    conda:        4.3.30-py36h5d9f9f4_0 --> 4.3.31-py36_0    
    conda-env:    2.6.0-0               --> 2.6.0-h36134e3_1 
    python:       3.6.2-0               --> 3.6.2-h02fb82a_12
    readline:     6.2-2                 --> 7.0-ha6073c6_4   
    sqlite:       3.13.0-0              --> 3.20.1-hb898158_2
    tk:           8.5.18-0              --> 8.6.7-hc745277_3 
    xz:           5.2.2-1               --> 5.2.3-h55aa19d_2 
[user@localhost x86_64-linux_CentOS-5]$ python -v
python: /lib64/libc.so.6: version `GLIBC_2.7' not found (required by python)
python: /lib64/libc.so.6: version `GLIBC_2.9' not found (required by python)
python: /lib64/libc.so.6: version `GLIBC_2.6' not found (required by python)

```

jeanconn commented:
Relatedly, it doesn't look to me like you can pin a *build*, and since there are Python 3.6.2 builds that work and don't work, we'd need to be careful.

jeanconn commented on cfg/anaconda-64.cfg:
```diff
@@ -10,6 +10,10 @@ cmds :
+    cp ${installer_dir}/pkgs.pinned ${prefix_arch}/conda-meta/pinned
+    cp ${installer_dir}/dot_condarc ${prefix_arch}/.condarc
+    # Python seems to need to be installed on its own to update
+    ${prefix_arch}/bin/conda install --quiet --yes --no-update-dependencies python=3.6.2_0
```
If we want to build the modules using the newest Python that seems to work, this seems to do it.  If we don't care about which Python was used in the build, this could be a post-facto update to the conda env.

taldcroft replied:
Since we have significant uncertainties about the process and potential gotchas, I think we want to be conservative and don't do anything funny like switching Python after the build.  I.e. we *do* care about the Python that was used for building.

taldcroft commented:
>  The fundamental problem is that generally we have no real insight into what built the package until something doesn't work.

I thought there was some identifier or somesuch that identified a package build as being in the "new" build system that is associated with CentOS-6. 

> I was hoping that we just didn't have to care if some module and ska_testr tests passed.

Ska_testr still has quite low coverage on the whole.  For instance we don't test numpy or scipy or matplotlib (or any other conda-installed packages) since we have (in the past) generally assumed those are going to work.  If we are using a CentOS-6 build for anything, I would say that all bets are off.  We may not have noticed any of this if we didn't happen to have a numba dependency in chandra_aca.

jeanconn commented:
> I thought there was some identifier or somesuch that identified a package build as being in the "new" build system that is associated with CentOS-6.

New packages built on the new compilers and new systems that *also* use the new conda should:
   - be in the new channel
   - have new hashes in the build string
   - have dependencies on new-style gcc  (libgcc-ng)
   - have a "burned in" recipe that references new compilers

However, none of these things were true of the llvmlite 0.20.0 package, so I don't have much confidence in the checking.

> If we are using a CentOS-6 build for anything

I agree, so I'd actually be inclined to just update chimchim.  But, since so far we can be confidence that broken packages will not have any *logical* failings, I think we can proceed with an update to ska3/flight and just revert if we notice any problems.

jeanconn commented:
> The latest estimate for updating chimchim is summer 2018.

Oh.  OK.  I think I missed that.  I thought we still had that in our pocket for a shorter-term update if things got unpleasant.  

> So let's go with what you have (python 3.6.2_0 in free packages) 

I tested numpy, scipy, numba on the vm and will retest and do matplotlib in a new build in /proj/sot/ska3/test_centos5 .

jeanconn commented:
Though it occurs to me that I don't actually know how best to review failures in the big packages.

taldcroft commented:
Is that a hypothetical or real concern?

jeanconn commented:
Not hypothetical but also probably not a real concern.  I get these failures running the scipy tests in my new environment, but it looks like they are also present in current ska3/flight 
```
FAILED (KNOWNFAIL=92, SKIP=1821, failures=2)
on
FAIL: test_decomp_update.TestQRdelete_F.test_delete_last_p_col
FAIL: test_iterative.test_convergence(<function cgs at 0x2b99a44a8c80>, <rand-cmplx-sym-pd>)
```

Also, not sure how to run the matplotlib tests from a conda package install:

```unagi: python -c "import matplotlib; matplotlib.test()"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/proj/sot/ska3/test_centos5/arch/x86_64-linux_CentOS-5/lib/python3.6/site-packages/matplotlib/__init__.py", line 1600, in test
    _init_tests()
  File "/proj/sot/ska3/test_centos5/arch/x86_64-linux_CentOS-5/lib/python3.6/site-packages/matplotlib/__init__.py", line 1558, in _init_tests
    raise ImportError("matplotlib test data is not installed")
ImportError: matplotlib test data is not installed
```
as I'm not sure if the test data needs to match the matplotlib version:
```unagi: conda install matplotlib-tests
Fetching package metadata ...........
Solving package specifications: .

Package plan for installation in environment /proj/sot/ska3/test_centos5/arch/x86_64-linux_CentOS-5:

The following NEW packages will be INSTALLED:

    matplotlib-tests: 1.5.1-py_0 defaults

Proceed ([y]/n)? n
```

jeanconn commented:
fido ska_testr output from test_centos5 environment (this PR):

```
**************************************************************
***      Package                  Script            Status ***
*** ------------------ ---------------------------- ------ ***
***   Chandra.Maneuver                 test_unit.py   pass ***
***   Chandra.Maneuver           post_check_logs.py   pass ***
***       Chandra.Time                 test_unit.py   pass ***
***       Chandra.Time           post_check_logs.py   pass ***
*** Chandra.cmd_states                 test_unit.py   pass ***
*** Chandra.cmd_states           post_check_logs.py   pass ***
***         Quaternion                 test_unit.py   pass ***
***         Quaternion           post_check_logs.py   pass ***
***            Ska.DBI                 test_unit.py   pass ***
***            Ska.DBI           post_check_logs.py   pass ***
***          Ska.Numpy                 test_unit.py   pass ***
***          Ska.Numpy           post_check_logs.py   pass ***
***        Ska.ParseCM             test_unit_git.sh   pass ***
***        Ska.ParseCM           post_check_logs.py   pass ***
***          Ska.Shell                 test_unit.py   pass ***
***          Ska.Shell           post_check_logs.py   pass ***
***          Ska.Table             test_unit_git.sh   ---- ***
***          Ska.Table           post_check_logs.py   ---- ***
***     Ska.engarchive    test_make_archive_long.sh   ---- ***
***     Ska.engarchive                 test_unit.py   pass ***
***     Ska.engarchive           post_check_logs.py   pass ***
***            Ska.ftp                 test_unit.py   pass ***
***            Ska.ftp           post_check_logs.py   pass ***
***       Ska.quatutil             test_unit_git.sh   pass ***
***       Ska.quatutil           post_check_logs.py   pass ***
***            Ska.tdb                 test_unit.py   pass ***
***            Ska.tdb           post_check_logs.py   pass ***
***       acisfp_check              test_regress.sh   ---- ***
***       acisfp_check           post_check_logs.py   ---- ***
***       acisfp_check              post_regress.py   ---- ***
***              agasc                 test_unit.py   pass ***
***              agasc           post_check_logs.py   pass ***
***        chandra_aca                 test_unit.py   pass ***
***        chandra_aca           post_check_logs.py   pass ***
***            cxotime                 test_unit.py   pass ***
***            cxotime           post_check_logs.py   pass ***
***          dea_check              test_regress.sh   ---- ***
***          dea_check           post_check_logs.py   ---- ***
***          dea_check              post_regress.py   ---- ***
***          dpa_check              test_regress.sh   ---- ***
***          dpa_check           post_check_logs.py   ---- ***
***          dpa_check              post_regress.py   ---- ***
***               kadi         test_regress_long.sh   ---- ***
***               kadi                 test_unit.py   pass ***
***               kadi           post_check_logs.py   pass ***
***               kadi         post_regress_long.py   ---- ***
***              maude                 test_unit.py   pass ***
***              maude           post_check_logs.py   pass ***
***               mica                 test_unit.py   pass ***
***               mica           post_check_logs.py   pass ***
***   package_manifest              test_regress.sh   pass ***
***   package_manifest           post_check_logs.py   pass ***
***   package_manifest              post_regress.py   pass ***
***           parse_cm                 test_unit.py   pass ***
***           parse_cm           post_check_logs.py   pass ***
***             pyyaks                 test_unit.py   pass ***
***             pyyaks           post_check_logs.py   pass ***
***          starcheck              test_regress.sh   ---- ***
***          starcheck           post_check_logs.py   ---- ***
***          starcheck              post_regress.py   ---- ***
***          timelines test_unit_git_sybase_long.sh   ---- ***
***          timelines           post_check_logs.py   ---- ***
***               xija                 test_unit.py   pass ***
***               xija           post_check_logs.py   pass ***
**************************************************************
```
using this test-spec
```
# Excludes
- find_attitude
- dpa_check
- dea_check
- acisfp_check
- kadi*long*
- Ska.engarchive*long*
- timelines
- starcheck
- Ska.Table
```

jeanconn commented:
@taldcroft I'm still not sure what to do with this.

taldcroft commented:
It's still on my list.

jeanconn commented on pkgs.conda:
```diff
@@ -70,3 +73,4 @@ setuptools
 conda
 conda-env
 pip
+sherpa=4.9.1
```
Sure, we could pin this.  I didn't think it *needed* to be pinned because it isn't a requirement for anything else (so nothing else is going to try to increment its version).  But it can't hurt to be explicit.

jeanconn commented on dot_condarc:
```diff
@@ -0,0 +1,8 @@
+channels:
+ - defaults
+ - cxc/channel/dev
```
And yes, sure we can remove the dev channel.  I just am not sure about the best way to document the sherpa conda package source (don't know if the one we used will stay in their dev channel anyway).  I also copied the sherpa conda package into /proj/sot/ska/pkgs so we'd have a spare (but conda packages really don't belong there).

taldcroft commented:
Ooops, we forgot to talk about this.  I had a concern this morning that by not pinning most packages, we run into a problem where a newly-installed dev ska has packages that are very different from flight ska (e.g. a new numpy release).  What's the current status (from the `conda install` perspective) of having every package pinned?  Does it work OK?

taldcroft commented:
And I just realized this is off-topic for this particular PR.

jeanconn commented:
For process, some discussion of this probably does belong in this particular PR.  Right now, by pinning llvmlite and reducing the set of repositories to the ones that are nominally CentOS5 compatible, there is the *hope* that:

   - a newly constructed Py3 skare will have a reasonable set of packages even if they are at higher versions that the ones in there now
   - an update of a set of packages will still be reasonable

If we need more security, I think we can just pin everything.  I see no reason why that wouldn't work from a `conda install` perspective, but I can certainly give it a go.  

A subtlety we may need to keep in mind is that pinning doesn't guaranteed that there isn't "build" creep at the same package version.

jeanconn commented:
Also, regarding promotion.  I built a candidate on a VM and put it in ska3/flight_to_promote as the aca user, with a plan to move ska3/flight and move or link this directory as flight, and then do a "make python_modules" to update Python packages that have been updated since I made the build.  Should be easy to back out if there are any problems.

taldcroft commented:
Despite a really strong desire to keep things simple and minimal, I do worry about assuming the package set will be "reasonable".  For a dev ska being used to test something on the side, what is important is that it is the same as flight.

Also, I hope we are designing this system to work beyond the window when we are stuck on CentOS-5, so we can't rely on the build system on the other end being basically frozen (by virtue of our channel selection).

Finally, it occurred to me that available package versions might be different on Windows or Mac.  This might be an entirely different headache, but it is at least good to know.

One possibility is to have regular major version updates of Skare where we unpin most everything, let the rest float up as possible, and then thoroughly re-test.  Could be every 6 months after the astropy release, but could be more frequent if needed.  This would apply only to external conda packages.

jeanconn commented:
> Also, I hope we are designing this system to work beyond the window when we are stuck on CentOS-5, so we can't rely on the build system on the other end being basically frozen (by virtue of our channel selection).

In general, I was trying not to argue with your strategy in this branch of basically just letting conda packages float where we can, because I thought you had a strong driver for that.  If there isn't a driver for that, we can certainly lock more stuff.  But regarding

> One possibility is to have regular major version updates of Skare where we unpin most everything, let the rest float up as possible, and then thoroughly re-test 

it seems then we need to track two different kinds of pins; "default" pins and explicit pins that we put there for some good reason.

jeanconn commented:
And I still haven't really figured out when we want to "pin" and when we just want to install at a specific version.

jeanconn commented:
Maybe we'll be fine if we leave pins for just the things that we're stuck with (right now llvmlite and sherpa).  

To set everything else, we'd just install from the environment yaml file, which we could add to the skare git repo at this point.  I think we had issues with that in older versions of conda but should be fine now.

jeanconn commented:
I actually had a go at trying to install from the environment yaml file, and I can't find a way to have that just work into the "root" of the new conda install.  I'm probably missing an option, but in the end we might be better off changing
```
    ${prefix_arch}/bin/conda-env export --name root > ${prefix_arch}/conda_env.yml
```
to 
```
${prefix_arch}/bin/conda list -e > ${prefix_arch}/versioned_pkgs.conda
```
(I think it used to be that anyway).  Then we can decide if we want to keep that versioned file also in skare version control.  To install the package list with all versions specified, you'd overwrite pkgs.conda with the versioned ... version.

taldcroft commented:
It seems like the file installed to `${prefix_arch}` can just be `pkgs.conda`.  See below.

I think we might just need an out-of-conda file that describes which conda packages absolutely need to be held at a certain version (llvmlite being the poster child).  Then the skare versioned `pkgs.conda` is originally created with the output of `conda list -e`.  For skare major releases, we temporarily edit `pkgs.conda` to unversion everything except llvmite etc, make the new environment, do full regression testing then re-do the `conda list -e`.  I think this will work and give full visibility into what happened to the environment.  And this only needs to be done for "major releases" i.e once every 6 months or so.

One concern is about version availability on different platforms (in particular Windows).  But right now Mark has his own build process anyway, so the releases `pkgs.conda` will document precisely the linux packages, and he can do what he does.

jeanconn commented:
> I think we might just need an out-of-conda file that describes which conda packages absolutely need to be held at a certain version (llvmlite being the poster child). 

Why do you think we need an out-of-conda file for this versus  the `pinned` file?

> For skare major releases, we temporarily edit pkgs.conda to unversion everything 

And what is your strategy for that? A script to chop off the versions?

There is one advantage of keeping a separate file recording the list of unversioned packages we *want* (basically the current pkgs.conda); sometimes the underlying dependences are actually simplified/removed.  Not that big a deal, but to me it seems conceptually simpler to actually *have* a list of just the packages/features we care about.

taldcroft commented:
Let's just start with some requirements and see if this is workable.

- No conda packages shall change version prior to a major release of Skare.
- When creating a fresh Ska environment, all conda packages shall have exactly the same version as  the corresponding major release.

By "conda packages" I mean external packages that are picked up in the conda channels you provide here.

My thinking has indeed evolved and I've gotten a little less comfortable in having no control over secondary dependencies.  Basically we install a lot of dependencies and I don't know we can say with confidence whether updates to the "obscure" ones might cause a problem.  But the converse is true, and we *can* say that if they don't float, then things should not break unexpectedly.

So I agree with the idea of having a list of unversioned packages that we really care about, with minimal explicit specification of downstream dependencies.  But this should only be used to update the complete versioned list at the time of a major Ska version update.

Whether we keep llvmlite at the desired version with pinning or elsewhere is an implementation detail.  I sort of think of the pinned file as a way to allow in-place updates to be pinned rather than fresh installs to be pinned, but I'm not adamant about that.

If you agree with that, ideas for implementation please!  :smile:

jeanconn commented:
I suppose I wasn't sure what to do here.  The PR is approved but then we started discussing new requirements to keep things working/safe in the future and you ended by asked for an implementation plan.  I think the implementation would still be what was was discussed earlier in the thread:

1) a pinned file with packages that have hard version requirements.  The pinning would help us to not accidentally update the environment in place in a way that would break these requirements.
2) A plan to keep our current pkgs.conda somewhere as the "unversioned" list of packages that we'd like in the the environment.  Only to be used when a many-package major update is desired.
3) A plan to make a strictly versioned pkgs.conda used as the default reference for new environments.

The naming conventions on the pkg files are TBD in that plan.  

So what needs to happen in general and what needs to happen in this PR to start using a Python 3 environment with the sherpa fixed?

taldcroft commented on pkgs.conda:
```diff
@@ -70,3 +73,5 @@ setuptools
 conda-env
 pip
+# Sherpa is also pinned in pkgs.pinned
+sherpa=4.9.1
```
First, shouldn't this only be in the `pkgs.pinned` like `llvmlite`?

But more to the point, why is sherpa being pinned?  Is 4.9.2 no good?

jeanconn replied:
Sure, I'll cut the version from the pkgs.conda.  I don't think it actually matters but it is better not to have it in more than one place if one place is sufficient.

> But more to the point, why is sherpa being pinned? Is 4.9.2 no good?

My impression was that Omar made a custom 4.9.1 for us on-the-side on CentOS 5, but they didn't revise their build procedure to go back to CentOS5 in general.  So, I figure any newer packages are not expected to work for us.

taldcroft replied:
Ah, well then pinning to 4.9.1 makes perfect sense.

jeanconn replied:
For sherpa, I wasn't sure if this PR should fetch 4.9.1 from the dev channel or just conda install the package that I fetched and saved in /proj/sot/ska/pkgs .

taldcroft replied:
Dev channel if that works.  Did I mention that JWST will be installing Ska3 soon?  They probably won't need sherpa, but we should be thinking in this direction.

jeanconn replied:
OK.  Thought I'm not sure how the JWST comment applies; did you give them a tar of packages or are they making their own or?

jeanconn commented:
OK, removed the explicit sherpa version in pkgs.conda, opened a new issue for the plan to control versioning more tightly in py3, and rebased.  I think /proj/sot/ska3/flight_to_promote is still ready-to-go at any time, though it will need a "make python_modules" to catch up on the changes since it was built in December.

jeanconn commented:
OK.  Does this need a chandracode warn/announce ?

taldcroft commented:
Yes on the chandracode warn and announce before installing.  A manifest of diffs in packages would be good.
